flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/committer/SinkCommitter.java

```diff
@@ -0,0 +1,463 @@
+          return 0;
+        }
+
+        @Override
+        public void signalFailedWithKnownReason(Throwable t) {}
+
+        @Override
+        public void signalFailedWithUnknownReason(Throwable t) {}
+
+        @Override
+        public void retryLater() {}
+
+        @Override
+        public void updateAndRetryLater(SinkCommittable committable) {}
+
+        @Override
+        public void signalAlreadyCommitted() {}
+      };
+
+  @VisibleForTesting
+  static final String MAX_CONTINUOUS_EMPTY_COMMITS = "flink.max-continuous-empty-commits";
+
+  public static final long INITIAL_CHECKPOINT_ID = -1L;
+  public static final String FLINK_JOB_ID = "flink.job-id";
+  public static final String OPERATOR_ID = "flink.operator-id";
+  private final String branch;
+  private final Map<String, String> snapshotProperties;
+  private final boolean replacePartitions;
+  private final int workerPoolSize;
+  private final String prefix;
+  private transient IcebergFilesCommitterMetrics committerMetrics;
+  private transient Table table;
+  private transient int maxContinuousEmptyCommits;
+  private transient ExecutorService workerPool;
+  private transient int continuousEmptyCheckpoints = 0;
+
+  public SinkCommitter(
+      TableLoader tableLoader,
+      String branch,
+      Map<String, String> snapshotProperties,
+      boolean replacePartitions,
+      int workerPoolSize,
+      String prefix,
+      IcebergFilesCommitterMetrics committerMetrics) {
+    this.branch = branch;
+    this.snapshotProperties = snapshotProperties;
+    this.replacePartitions = replacePartitions;
+    this.workerPoolSize = workerPoolSize;
+    this.prefix = prefix;
+    this.committerMetrics = committerMetrics;
+
+    if (!tableLoader.isOpen()) {
+      tableLoader.open();
+    }
+
+    this.table = tableLoader.loadTable();
+    this.maxContinuousEmptyCommits =
+        PropertyUtil.propertyAsInt(table.properties(), MAX_CONTINUOUS_EMPTY_COMMITS, 10);
+    Preconditions.checkArgument(
+        maxContinuousEmptyCommits > 0, MAX_CONTINUOUS_EMPTY_COMMITS + " must be positive");
+    this.workerPool =
+        ThreadPools.newWorkerPool("iceberg-worker-pool-" + table + "-" + prefix, workerPoolSize);
+    this.continuousEmptyCheckpoints = 0;
+  }
+
+  @Override
+  public void commit(Collection<CommitRequest<SinkCommittable>> commitRequests)
+      throws IOException, InterruptedException {
+    NavigableMap<Long, byte[]> manifestMap = Maps.newTreeMap();
+    NavigableMap<Long, CommitRequest<SinkCommittable>> commitRequestMap = Maps.newTreeMap();
+
+    if (commitRequests.isEmpty()) {
+      commitRequests.add(EMPTY_COMMIT_REQUEST);
+    }
+
+    for (CommitRequest<SinkCommittable> request : commitRequests) {
+      manifestMap.put(request.getCommittable().checkpointId(), request.getCommittable().manifest());
+      commitRequestMap.put(request.getCommittable().checkpointId(), request);
+    }
+
+    long maxCommittedCheckpointId = getMaxCommittedCheckpointId(commitRequestMap);
```

Comment:
this involves a check/read of the table snapshot history for every commit, which add to latency. the behavior/overhead is different. Current code only retrieve the max committed checkpoint id during job initialization phase.

Reply:
This is needed, so that this class can keep up with the design contract of `Comitter`:
```
A commit must be idempotent: If some failure occurs in Flink during commit phase, Flink will restart from previous checkpoint and re-attempt to commit all committables. 
```

Reply:
I understand this is needed. it is unfortunate that the new sink V2 API enforce unnecessary check/overhead to Iceberg sink. we don't have to resolve it in this PR. but it is sth worth follow up to see if we can avoid it.

New sink V2 tried to hide a lot things from connectors, which also result in this side effect that connectors lose control.